src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp

```diff
@@ -0,0 +1,440 @@
+    }
+    SessionContext *sessionCtxt = ue_ctxt->getSessionContext();
+    if (sessionCtxt == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "send_mb_req_to_sgw_for_ho: session ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    BearerContext *bearerCtxt = sessionCtxt->getBearerContext();
+    if (bearerCtxt == NULL)
+    {
+        log_msg(LOG_DEBUG, "send_mb_req_to_sgw_for_ho: bearer ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+
+    struct MB_Q_msg mb_msg;
+    MmeGtpMsgUtils::populateModifyBearerRequestHo(cb, *ue_ctxt,
+    		*ho_ctxt, mb_msg);
+    if (bearerCtxt)
+        bearerCtxt->setS1uEnbUserFteid(Fteid(mb_msg.s1u_enb_fteid));
+
+    cmn::ipc::IpcAddress destAddr;
+    destAddr.u32 = TipcServiceInstance::s11AppInstanceNum_c;
+
+    mmeIpcIf_g->dispatchIpcMsg((char*) &mb_msg, sizeof(mb_msg), destAddr);
+
+    ProcedureStats::num_of_mb_req_to_sgw_sent++;
+    log_msg(LOG_DEBUG, "Leaving send_mb_req_to_sgw_for_ho \n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : process_mb_resp_for_ho
+***************************************/
+ActStatus ActionHandlers::process_mb_resp_for_ho(ControlBlock& cb)
+{
+    log_msg(LOG_DEBUG, "Inside process_mb_resp_for_ho \n");
+
+    ProcedureStats::num_of_processed_mb_resp ++;
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : send_s1_rel_cmd_to_ue_for_ho
+***************************************/
+ActStatus ActionHandlers::send_s1_rel_cmd_to_ue_for_ho(ControlBlock& cb)
+{
+    log_msg(LOG_DEBUG, "Inside send_s1_rel_cmd_to_ue_for_ho\n");
+    UEContext *ue_ctxt = static_cast<UEContext*>(cb.getPermDataBlock());
+    if(ue_ctxt == NULL)
+    {
+    	log_msg(LOG_DEBUG, "send_s1_rel_cmd_to_ue_for_ho: ue context is NULL \n");
+    	return ActStatus::HALT;
+    }
+
+    S1HandoverProcedureContext *ho_ctxt =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (ho_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG, "process_ho_notify: procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+    struct s1relcmd_info s1relcmd;
+    s1relcmd.msg_type = s1_release_command;
+    s1relcmd.ue_idx = ue_ctxt->getContextID();
+    s1relcmd.enb_s1ap_ue_id = ho_ctxt->getSrcS1apEnbUeId();
+    s1relcmd.cause.present = s1apCause_PR_radioNetwork;
+    s1relcmd.cause.choice.radioNetwork = s1apCauseRadioNetwork_successful_handover;
+    s1relcmd.enb_fd = ho_ctxt->getSrcEnbFd();
+    /*Send message to S1AP-APP*/
+    cmn::ipc::IpcAddress destAddr;
+    destAddr.u32 = TipcServiceInstance::s1apAppInstanceNum_c;
+
+    mmeIpcIf_g->dispatchIpcMsg((char *) &s1relcmd, sizeof(s1relcmd), destAddr);
+
+    ProcedureStats::num_of_s1_rel_cmd_sent ++;
+    log_msg(LOG_DEBUG,"Leaving send_s1ap_ue_ctxt_rel_command_to_src_enb \n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : process_ue_ctxt_rel_comp_for_ho
```

Comment:
should update the ENB FD stored in UE ctxt here. It should be updated to the target ENB cbIndex.

Reply:
Its worth to do following testing after handover completion
1. data testing in uplink/down link direction just to make sure all tunnel points are correctly set
2. Control channel testing - by doing disconnect from UE..see if we have any issue in sending back response to new eNB,..

Reply:
> should update the ENB FD stored in UE ctxt here. It should be updated to the target ENB cbIndex.

Target eNB FD, enb-ue-s1ap-id is getting updated to UE context when we process HO Notify.